source/hpack/encoder.d

```diff
@@ -0,0 +1,265 @@
+module hpack.encoder;
+
+import hpack.tables;
+import hpack.huffman;
+import hpack.util;
+import hpack.exception;
+
+import vibe.http.status;
+import vibe.http.common;
+
+import std.range;
+import std.bitmanip; // prefix encoding / decoding
+import std.typecons;
+import std.conv;
+import std.exception;
+import std.array;
+
+struct HeaderEncoder(T = HTTP2HeaderTableField[])
+	if(isInputRange!T && is(ElementType!T : HTTP2HeaderTableField) ||
+	   is(T == HTTP2HeaderTableField))
+{
+	private {
+		static if(is(T == HTTP2HeaderTableField)) T[] m_range;
+		else T m_range;
+		IndexingTable m_table;
+		ubyte[] m_encoded;
+	}
+
+	// flag to select huffman encoding / raw encoding
+	bool huffman;
+
+	this(T range, IndexingTable table, bool huff = true) @trusted
+	{
+		static if(is(T == HTTP2HeaderTableField)) m_range ~= range;
+		else m_range = range;
+
+		m_table = table;
+		huffman = huff;
+
+		encode();
+	}
+
+// InputRange specific methods
+	@property bool empty() @safe @nogc { return m_encoded.empty; }
+
+	@property ubyte[] front() @safe @nogc { return m_encoded; }
+
+	@property ubyte[] back() @safe @nogc { return m_encoded; }
+
+	void popFront() @trusted
+	{
+		enforce!HPACKException(!empty, "Cannot call popFront on an empty HeaderDecoder");
+
+		m_encoded.popFront();
```

Comment:
Since `.front` is of type `ubyte[]`, assuming the input range interface is kept, `.popFront` should just set `m_encoded = null;` instead of popping the first byte. This potential bug is currently apparently hidden by the following call to `encode()`. More idiomatic would be to let `front` return a `ubyte` and call `encode` only if `m_encoded` is empty (this may be slower if the inliner doesn't work optimally, but usually should be as fast).